Advanced pickit

RB-Development opened the pull request:
**To use the advanced itemfilter an ID-Tome needs to be in the inventory.**

This PR allows you to decide if magic, rare and uniques should be stashed or not.
Furthermore you are now able to specify nested filters in your pickit.ini
(Examples are in pickit.ini)

Added several item properties in the assets folder.

aeon0 commented:
GJ on the effort. I am having a few conceptional issues with it though. There is already an "id items at cain" feature. Imho this is a lot more "safe" and reliable. We all know the issues we can face when doing stuff in the inventory...
Also I am not a big fan of the yml file which adds yet another layer of pickit info to a different place in a different format. Why cant we add these values just like we do with the socket info or etheral info in the pickit.ini?

RB-Development commented:
> 

Havent mentioned the "id @ Cain" feature. I will give it a try and implement the routine instead of the ID-Tome.
Just need to implement some logic while picking up items if an identification "@Cain" ist needed.
Ill try to make the attributes availible in the pickit.ini file so we dont need a separate file. Hope it wont get to messy :)

RB-Development commented:
Integrated yaml informationen into pickit.ini. No yaml file needed.
Integrated routine to check if identification is activated else the include for magic, rare, sets and uniques are disabled.
Only donwside: no nested attributation in pickit.ini possible

aeon0 commented on config/pickit.ini:
```diff
@@ -177,9 +176,9 @@ magic_unearthed_wand=0
+magic_greater_talons=1, and(2_assasin_skills, 3_lightning_sentry)
 magic_feral_claws=0
-magic_runic_talons=0
+magic_runic_talons=1, and(2_assasin_skills, 3_lightning_sentry)
```
So we dont allow +3 trap, +3 lightning? Maybe only filter for +3_lightning?

RB-Development replied:
changed to pickup 2 assa, 3 Lightning and traps+ lightning sentry.
((2_assasin_skills, 3_lightning_sentry), (traps, lightning_sentry))

aeon0 commented on config/pickit.ini:
```diff
@@ -207,15 +206,15 @@ magic_ogre_gauntlets=0
+rare_amulet=1, (2_amazon_skills, 2_all_skills, 2_assasin_skills, 2_barbarian_skills, 2_Druid_skills, 2_necromancer_skills, 2_paladin_skills, 2_sorceress_skills)
 rare_jewel=1
-rare_ring=1
+rare_ring=1, (and(life_stolen, mana_stolen), faster_cast_rate)
```
puh. There can be really good stats rings without fcr or ll, ml. Rings don't take up much space, I'd vote to just keep them all.

RB-Development replied:
Removed that include filter in this case

aeon0 commented on config/pickit.ini:
```diff
@@ -207,15 +206,15 @@ magic_ogre_gauntlets=0
-rare_coronet=1
-rare_diadem=1
-rare_tiara=1
+rare_circlet=1, ((faster_cast_rate, 2_amazon_skills), (faster_cast_rate, 2_assasin_skills), (faster_cast_rate, 2_barbarian_skills), (faster_cast_rate, 2_Druid_skills), (faster_cast_rate, 2_necromancer_skills), (faster_cast_rate, 2_paladin_skills), (faster_cast_rate, 2_sorceress_skills), (2_all_skills, faster_cast_rate), (3_pcombat, faster_cast_rate), (faster_run_walk, 2_amazon_skills), (faster_ru ...
```
Can rare circlets get 3_pcombat? I thought only magic can.

RB-Development replied:
The cant. Removed that part

aeon0 commented on src/bot.py:
```diff
@@ -108,7 +108,7 @@ def __init__(self, screen: Screen, game_stats: GameStats, pick_corpse: bool = Fa
         self._pausing = False
         self._current_threads = []
-        self._no_stash_counter = 0
+        self._no_stash_counter = 5
```
Any reason to set this to 5?

RB-Development replied:
Just for Debug to trigger stashing when bot start. Set to 0 again

aeon0 reviewed (changes requested):
And obviously a bit of clean up, removing yml and commented code. But its going in the right direction. Good job.

aeon0 reviewed (changes requested):
I will wait with merging for next release because I would see this PR as super nice, but also somewhat risky to have unexpected behaviours. So the course of actions would be, do a new release, merge this and test it properly on master and get ready for the next one.

RB-Development commented on src/ui/ui_manager.py:
```diff
@@ -268,6 +268,11 @@ def _keep_item(self, item_finder: ItemFinder, img: np.ndarray) -> bool:
+            if ("magic_small_charm" in x.name):
+                print ("just for DEBUG")
+            #Disable include params for uniq, rare, magical if ident is disabled in params.ini
+            if (not self._config.char["id_items"]) and ("uniq" in x.name or "magic" in x.name or "rare" in x.name or "set" in x.name):
```
Y, working fine. tested :)

RB-Development commented:
> I will wait with merging for next release because I would see this PR as super nice, but also somewhat risky to have unexpected behaviours. So the course of actions would be, do a new release, merge this and test it properly on master and get ready for the next one.

Totally agree. Im also working on a gambling option. Right now an alpha version is running but needs some small improvements before i can start another PR.

aeon0 commented on config/pickit.ini:
```diff
@@ -43,7 +42,7 @@ misc_greater_healing_potion=0
 
 ;misc
-misc_gold=0
+misc_gold=1
```
If you make misc_gold default 1 I would also make gold_trav_only param in [car] section default 1

aeon0 reviewed (changes requested):
I wouldn't give the code any elegance award, kind of brute force, but I'd be ok with it given the fact that we plan to rewrite most of the item pickit with ocr and a more detailed item description format.
But one thing I would ask for are unittests for the parse function. It seems like code that could 1) be easily unittested 2) would be very beneficial for it.

RB-Development commented:
Some small changes made to make unittests comftable. Unittests added, small adjustment regarding pickit

aeon0 commented:
@RB-Development In an attempt to solve the merge conflict on the github GUI I think I broke sth maybe? Can you check? Otherwise I'd fine merging.

RB-Development commented:
Looks like you killed the function in config.py to read the nested filters
`def parse_item_config_string(self, key: str = None) -> ItemProps:
        string = self._select_val("items", key).upper()
        return self.string_to_item_prop (string)

    def string_to_item_prop (self, string: str) -> ItemProps:
        item_props = ItemProps()
        brk_on = 0
        brk_off = 0
        section = 0
        counter = 0
        start_section = 0
        start_item = 0
        include_list = []
        exclude_list = []
        for char in string:
            new_section = False
            counter+=1
            if char == "(":
                brk_on +=1
            elif char == ")":
                brk_off += 1
            if ((char == "," and brk_on==brk_off)):
                new_section = True
                if  (counter == len (string)):
                    string_section = string [start_section:counter]
                else:
                    string_section = string [start_section:counter-1]
                if section == 0:
                    item_props.pickit_type = int (string_section)
                section +=1
                start_section = counter
            if ((char == "," and (brk_on==brk_off+1)) or new_section or counter == len (string)):
                if new_section: 
                    section -=1
                if start_item ==0:
                    start_item = start_section
                if  (counter == len (string)):
                    item = string [start_item:counter]
                else:
                    item = string [start_item:counter-1]
                if section == 0 and counter == len (string):
                    item_props.pickit_type = int (item)
                    pass
                if section ==1:
                    include_list.append (item)
                    start_item = counter +1
                elif section ==2:
                    exclude_list.append (item)
                    start_item = counter +1
                if new_section: 
                    section +=1
        if (len (include_list)>0 and (len (include_list[0]) >=6)):
            if ("AND" in include_list[0][0: 6] and not ")" in include_list[0]):
                item_props.include_type = "AND"
            else:
                item_props.include_type = "OR"
        if (len (exclude_list)>0 and (len (exclude_list[0]) >=6)):
            if ("AND" in exclude_list[0][0:6] and not ")" in include_list[0]):
                item_props.exclude_type = "AND"
            else:
                item_props.exclude_type = "OR"
        for i in range (len(include_list)):
            include_list[i]  = include_list[i].replace (" ","").replace ("OR(","").replace ("AND(", "").replace ("(", "").replace (")","")
            include_list[i]  = include_list[i].split (",")
        for l in range (len (exclude_list)):
            exclude_list[l] = exclude_list[l].replace (" ","").replace ("OR(","").replace ("AND(", "").replace ("(", "").replace (")","")
            exclude_list[l] = exclude_list[l].split (",")
        item_props.include = include_list
        item_props.exclude = exclude_list
        return item_props`

RB-Development commented:
I added those missing function in config.py and made some bugfixing regarding to changing function to static_methods in config.py
Should be fine now.

RB-Development commented:
Hey @aeon0 anymore changes need to be done for merging?

Steve6030 commented:
Using this PR. Working well. Great job.

mgleed reviewed (changes requested):
@RB-Development Since aeon reviewed and gave it the go-ahead earlier, I'd be OK merging in his absence. Only thing I'd ask though before merge is this--will you write a better tutorial for how to implement this? Either modify top of pickit.ini further than you have, update your writeup from initial commit with PR, etc. Still a little unsure what I'm supposed to do once I've chosen to use advanced pickit.

RB-Development commented:
@mgleed added documentary in pickit.ini and changed the inital comment regarding the PR. Hope that matches your expectation :)

mgleed commented on src/ui/ui_manager.py:
```diff
@@ -269,6 +269,10 @@ def _keep_item(self, item_finder: ItemFinder, img: np.ndarray, do_logging: bool
             if ("potion" in x.name) or (self._config.items[x.name].pickit_type == 0): continue
             include_props = self._config.items[x.name].include
             exclude_props = self._config.items[x.name].exclude
+            #Disable include params for uniq, rare, magical if ident is disabled in params.ini
+            #if (not self._config.char["id_items"]) and ("uniq" in x.name or "magic" in x.name or "rare" in x.name or "set" in x.name):
+            if (not self._config.char["id_items"]) and any(item_type in x.name for item_type in ["uniq", "magic", "rare", "set"]):
+                include_props = False 
             if not (include_props or exclude_props):
```
Some people prefer not to identify items. If they have `id_items=0` shouldn't we set include *and* exclude props to none? Some properties like sockets won't show up on unidentified items iirc.

mgleed reviewed (changes requested):
Looks pretty good, and I tested and it works nicely without needing to adjust anything on my end. I think this could potentially be merged and enjoyed by the masses for the next "who knows how long" while we wait for OCR item filtering. I'd better leave it to @aeon0 to decide now because he's back.

RB-Development commented:
> Looks pretty good, and I tested and it works nicely. I think this could potentially be merged and enjoyed by the masses for the next "who knows how long" while we wait for OCR item filtering. I'd better leave it to @aeon0 to decide now because he's back.

Even with ocr this could be a benefit because you are able to use nested filters. We can throw away the assets part but the Rest should work even with ocr

RB-Development commented:
@mgleed Totally fine to keep gold options to default.
